dashmips/debuggerserver.py

```diff
@@ -65,23 +45,40 @@ def debug_mips(program: MipsProgram, host="localhost", port=2390, should_log=Fal
         format="%(asctime)-15s %(levelname)-7s %(message)s",
         level=log.INFO if should_log else log.CRITICAL,
     )
-    logger = log.getLogger("websockets.server")
+    logger = log.getLogger("sockets.server")
     logger.addHandler(log.StreamHandler())
-    log.info(f"Serving on: ws://{host}:{port}")
-
-    # Collect functions from debugger.py
-    debugger_module = importlib.import_module(".debugger", "dashmips")
-    funcs = inspect.getmembers(debugger_module, inspect.isfunction)
-    commands = {}
-    for name, command in funcs:
-        commands[name.replace("debug_", "")] = functools.partial(command, program=program)
-
-    ws_func = functools.partial(dashmips_debugger, commands=commands)
-    start_server = websockets.serve(ws_func, host, port, close_timeout=2000)
-    loop = asyncio.get_event_loop()
-    try:
-        ws_server = asyncio.get_event_loop().run_until_complete(start_server)
-        loop.run_forever()
-    except KeyboardInterrupt:
-        log.warning('Shutting down debugger...')
-        loop.close()
+    log.info(f"Serving on: s://{host}:{port}")
+
+    class DashmipsTCPServerHandler(socketserver.BaseRequestHandler):
+        def setup(self):
+            # Collect functions from debugger.py
+            debugger_module = importlib.import_module(".debugger", "dashmips")
+            funcs = inspect.getmembers(debugger_module, inspect.isfunction)
+            self.commands = {}
+            for name, command in funcs:
+                self.commands[name.replace("debug_", "")] = functools.partial(command, program=program)
+
+        def handle(self):
+            # self.request is the TCP socket connected to the client
+            while True:
+                header = b""
+                while True:
+                    header += self.request.recv(1)
+                    if header and chr(header[-1]) == '}':
+                        break
```

Comment:
I'm fine with this method, let us be a little more defensive and add a condition that if header is 1000 characters and we still haven't seen the closing curly we should exit with with failure.

Reply:
How would you prefer the program exit?  Should it throw an exception and log an error message?

Reply:
Good question, makes me realize we might have added a bit of a bug here with program exiting in general, the old client loop function used to break the websocket loop which implicitly would lead to an exit. My advice it to have one actual sys.exit() point in the program and to reach that point via exceptions. 

So I think `try: except:`-ing around `serve_forever` is appropriate, with except blocks for KeyboardInterruption, ProgramExit and CommunicationError (or ParseError, you decide) (the last two being exceptions of your own making) and handle them appropriately (for example program exit isn't actually a bad thing, we just want consistent exit strategies) 

You should log what you think is best so here's a note about levels:
Will only be seen if someone passed `-l` to dashmips:
- `debug` - for a developer to get detailed insight on the internals of a program (like dump of network traffic)
- `info` - verbose information, like entering certain blocks of code or calling functions, intended for someone debugging an issue in production
- `warning` - not an error but something that should be corrected

Will always print:
- `critical` or `error` - something unexpected occurred the program cannot continue, this is in the realm of socket port in use, or permission denied, things _entirely_ unfixable by our program

Reply:
I committed the changes I made, it should now exit with a message on both normal completion and if the message is too long (> 1000).  Let me know if you still want me to add an exit for `KeyboardInterruption`!